src/dune_engine/action_builder.ml

```diff
@@ -5,12 +5,13 @@ module T = struct
   type 'a t =
     | Pure : 'a -> 'a t
     | Map : ('a -> 'b) * 'a t -> 'b t
+    | Bind : 'a t * ('a -> 'b t) -> 'b t
     | Both : 'a t * 'b t -> ('a * 'b) t
     | Seq : unit t * 'b t -> 'b t
     | All : 'a t list -> 'a list t
     | Map2 : ('a -> 'b -> 'c) * 'a t * 'b t -> 'c t
-    | Paths_for_rule : Path.Set.t -> unit t
     | Paths_glob : File_selector.t -> Path.Set.t t
+    | Source_tree : Path.t -> Path.Set.t t
```

Comment:
Should the argument here be `Path.Source.t`?

Reply:
Probably. We need to turn the paths into ones in the current build context, so it would end up being `{ context_name : string; dir : Path.Source.t }`. This seems better, but let's leave that to another PR, for now I kept what we were doing.